# Working with Calendar Equations

## What are Calendar Equations?

Starting in the fourth century BCE Athenians, when recording decrees of the ekklesia, frequently used a standard preamble for inscriptions that gave both a prytany date and a festival calendar date. [IG II/III³ 1, 338](http://telota.bbaw.de/ig/digitale-edition/inschrift/IG%20II_III%C2%B3%201,%20338) (=IG II² 338) is a good example that offers the fullest form with no restorations:

    ἐπὶ Νικοκράτους ἄρχοντος, ἐπὶ τῆς Αἰγηίδος
    πρώτης πρυτανείας, ἧι Ἀρχέλας Χαιρίου Παλ–
    ληνεὺς ἐγραμμάτευεν· Μεταγειτνιῶνος ἐνά–
    τηι ἱσταμένου· ἐνάτηι καὶ τριακοστῆι τῆς
    πρυτανείας
    
The parts of this preamble are:

1. The árkhōn: ἐπὶ Νικοκράτους ἄρχοντος, "in the árkhōnship of Nikokrátēs..."
2. The prytany and tribe that held it: ἐπὶ τῆς Αἰγηίδος πρώτης πρυτανείας, "...in the first prytany, held by Aigēis..."
3. The secretary: ἧι Ἀρχέλας Χαιρίου Παλληνεὺς ἐγραμμάτευεν, "...for which Arkhélas son of Khairías from Pallēnē was secretary..."
4. The festival date: Μεταγειτνιῶνος ἐνάτηι ἱσταμένου, "...9th of Metageitniṓn..."
5. The prytany day: ἐνάτηι καὶ τριακοστῆι τῆς πρυτανείας, "...39th of the prytany"

These two dates provide a _calendar equation_: 

    Metageitniṓn 9 = Prytany I.39

In some cases, the equation provides useful data about the calendar for the year, or the calendar in general. For instance, this inscription tells us three things:

1. The year (333/332 BCE, which we know from the árkhōn) was **intercalary** because this prytany (the first) has 39 days. In a normal year, the longest prytanies were 36 days.
2. Hekatombaiṓn was **full**. If the 39th day of the prytany is Metageitniṓn 9, the 31st would be Metageitniṓn 1, and therefore the 30th has to be Hekatombaiṓn 30. 
3. The intercalated month was **not** Hekatombaiṓn because having a second Hekatombaiṓn before Metageitniṓn would push Metageitniṓn into the second prytany. This is hardly surprising, but it is worth keeping in mind.

I.39 is the 39th day of the year, as is Metageitniṓn 9 if Hekatombaiṓn is full (30 + 9 = 39). The solution to this equation, then, is

    Met 9 = I 39 = DOY 39

The year is intercalary, and begins with a full Hekatombaiṓn and 39-day prytany I.

## Using `heniautos` with Calendar Equations
With an equation such as this, `heniautos` can help determine the possible alignments of the two calendars, expressed as being the same DOY (day of the year). It is worth looking at an example with more possible solutions.

[IG II³ 345](http://telota.bbaw.de/ig/digitale-edition/inschrift/IG%20II_III%C2%B3%201,%20345) (= II² 345) is similarly complete (at least in terms of the dates), but comes much later in this year—332/1 BCE in this case..

    ἐπὶ Νικήτου ἄρ[χοντος, ἐ]πὶ τῆς Ἀντιοχ–
    ίδος ὀγδόης π[ρυτανεία]ς, ἧι Ἀριστόνο–
    υς Ἀριστόνο[υ Ἀναγυράσι]ος ἐγραμ[μά]τ–
    ευεν· Ἐλαφη[βολιῶνος ἐν]άτηι ἐπὶ δέκα·
    ἑβδόμ[ηι τῆς πρυτανείας]
    
What are the possible sollutions? for 

    Elaphēboliṓn 19 = Prytany VIII 7?
    
`heniautos` provides three functions to help with this: `festival_doy()`, `prytany_doy()`, and `equations()`

### `festival_doy()`

`festival_doy()` will calculate the possible days of the year for a date on the festival calendar:

In [2]:
import heniautos as ha
ha.init_data()

ha.festival_doy(ha.Months.ELA, 19)

({'date': (<Months.ELA: 9>, 19),
  'doy': 254,
  'preceding': (30, 30, 30, 29, 29, 29, 29, 29),
  'intercalation': False},
 {'date': (<Months.ELA: 9>, 19),
  'doy': 255,
  'preceding': (30, 30, 30, 30, 29, 29, 29, 29),
  'intercalation': False},
 {'date': (<Months.ELA: 9>, 19),
  'doy': 256,
  'preceding': (30, 30, 30, 30, 30, 29, 29, 29),
  'intercalation': False},
 {'date': (<Months.ELA: 9>, 19),
  'doy': 257,
  'preceding': (30, 30, 30, 30, 30, 30, 29, 29),
  'intercalation': False},
 {'date': (<Months.ELA: 9>, 19),
  'doy': 258,
  'preceding': (30, 30, 30, 30, 30, 30, 30, 29),
  'intercalation': False},
 {'date': (<Months.ELA: 9>, 19),
  'doy': 283,
  'preceding': (30, 30, 30, 29, 29, 29, 29, 29, 29),
  'intercalation': True},
 {'date': (<Months.ELA: 9>, 19),
  'doy': 284,
  'preceding': (30, 30, 30, 30, 29, 29, 29, 29, 29),
  'intercalation': True},
 {'date': (<Months.ELA: 9>, 19),
  'doy': 285,
  'preceding': (30, 30, 30, 30, 30, 29, 29, 29, 29),
  'intercalation': True},
 {'date

Each `dict` in these results represents one possibility for Elaphēboliṓn 19. The first possibility is:

    {'date': (<Months.ELA: 9>, 19),
     'doy': 254,
     'preceding': (30, 30, 30, 29, 29, 29, 29, 29),
     'intercalation': False}
     
This means that Elaphēboliṓn 19 can be DOY 254 if it is preceded by 3 full and five hollow months—the list of month lengths in `preceding`—and that none of them are intercalated months. Since Elaphēboliṓn is normally the 9th month, if it is preceded by eight months they are also all the normal months.

Note that the list of month lengths in `preceding` is a _count_ only. No order is implied and they could be rearranged in any way without changing the result that:

    (30 × 3) + (29 × 5) + 19 = 254
    
The last possibility is:

    {'date': (<Months.ELA: 9>, 19),
     'doy': 287,
     'preceding': (30, 30, 30, 30, 30, 30, 30, 29, 29),
     'intercalation': True})
     
This means that Elaphēboliṓn 19 can be DOY 287 if it is preceded by 7 full and 2 hollow months—nine in total meaning that this Elaphēboliṓn is the _tenth_ month and one of the preceding months is intercalated. Technically, it is also possible the the Elaphēboliṓn input to `festival_doy()` is the intercalated month. That is, the ninth month is normal Elaphēboliṓn and this tenth month is _Elaphēboliṓn hústeros_. We would expect that to be indicated in the inscription but it is not always so clear.

Altogether without a preceding intercalation Elaphēboliṓn 19 can fall in the range of DOY 254–258. With a preceding intercalation it can fall in the range DOY 283–287.

Since no year can have more than seven full months or six hollow months, `festival_doy()` does not return possibilites that exceed these. For instance, Elaphēboliṓn 19 _could_ be DOY 288 if it was preceded by eight full months and one hollow but eight full months in a year is impossible.

### `prytany_doy()`

`prytany_doy()` performs the equivalent calculation for the conciliar year. This function requires either a specific prytany type (a `Prytany` constant), or a year with `Prytany.AUTO`. These two invocations return the same results:

    ha.prytany_doy(ha.Prytanies.VIII, 7, pryt_type=ha.Prytany.ALIGNED_10)
    ha.prytany_doy(ha.Prytanies.VIII, 7, pryt_type=ha.Prytany.AUTO, 
        ha.bce_as_negative(333))

In [3]:
ha.prytany_doy(ha.Prytanies.VIII, 7, pryt_type=ha.Prytany.ALIGNED_10) 

({'date': (<Prytanies.VIII: 8>, 7),
  'doy': 253,
  'preceding': (36, 35, 35, 35, 35, 35, 35),
  'intercalation': False},
 {'date': (<Prytanies.VIII: 8>, 7),
  'doy': 254,
  'preceding': (36, 36, 35, 35, 35, 35, 35),
  'intercalation': False},
 {'date': (<Prytanies.VIII: 8>, 7),
  'doy': 255,
  'preceding': (36, 36, 36, 35, 35, 35, 35),
  'intercalation': False},
 {'date': (<Prytanies.VIII: 8>, 7),
  'doy': 256,
  'preceding': (36, 36, 36, 36, 35, 35, 35),
  'intercalation': False},
 {'date': (<Prytanies.VIII: 8>, 7),
  'doy': 274,
  'preceding': (39, 38, 38, 38, 38, 38, 38),
  'intercalation': True},
 {'date': (<Prytanies.VIII: 8>, 7),
  'doy': 275,
  'preceding': (39, 39, 38, 38, 38, 38, 38),
  'intercalation': True},
 {'date': (<Prytanies.VIII: 8>, 7),
  'doy': 276,
  'preceding': (39, 39, 39, 38, 38, 38, 38),
  'intercalation': True},
 {'date': (<Prytanies.VIII: 8>, 7),
  'doy': 277,
  'preceding': (39, 39, 39, 39, 38, 38, 38),
  'intercalation': True})

The results are similar to those from `festival_doy()`.

The first possibility is DOY 253:

    {'date': (<Prytanies.VIII: 8>, 7),
     'doy': 253,
     'preceding': (36, 35, 35, 35, 35, 35, 35),
     'intercalation': False}
     
`intercalation: False` has a slightly different meaning in the case of prytanies. Since whether a year is intercalary or not affects the _lengths_ of prytanies rather thatn their _number_, `intercalation: False` means the year must be intercalary, whether or not the intercalation falls before or after prytany VIII. DOY 253 is the result of 36- and 35-day prytanies, if the year is intercalary, they would be 39 and 38 days such as we have the DOY 277 solution

    {'date': (<Prytanies.VIII: 8>, 7),
     'doy': 277,
     'preceding': (39, 39, 39, 39, 38, 38, 38),
     'intercalation': True}
     
As with `festival_doy()`, the list of prytany lengths in `preceding` is a _count_ only. Each kind of conciliar year is made up of a certain number of long and short prytanies—such as four long and six short in 10-prytany years—`prytany_doy()` will not return impossible results.

Our results for prytany VIII 7 show that it can fall in the ranges DOY 253–256 (ordinary) or 274–277 (intercalary)

### `equations()`

A quick mental calculation will tell us that the two together are restricted to "solutions" in the range DOY 254-DOY 256, but `heniautos` will do the calculation for us with `equations()`. This function takes a tuple for the festival date, a tuple for the prytany date and, like `prytany_doy()` you must specify a prytany type or `Prytany.AUTO` with a year.

In [5]:
ha.equations((ha.Months.ELA, 19), (ha.Prytanies.VIII, 7), pryt_type=ha.Prytany.ALIGNED_10)

(({'date': (<Months.ELA: 9>, 19),
   'doy': 254,
   'preceding': (30, 30, 30, 29, 29, 29, 29, 29),
   'intercalation': False},
  {'date': (<Prytanies.VIII: 8>, 7),
   'doy': 254,
   'preceding': (36, 36, 35, 35, 35, 35, 35),
   'intercalation': False}),
 ({'date': (<Months.ELA: 9>, 19),
   'doy': 255,
   'preceding': (30, 30, 30, 30, 29, 29, 29, 29),
   'intercalation': False},
  {'date': (<Prytanies.VIII: 8>, 7),
   'doy': 255,
   'preceding': (36, 36, 36, 35, 35, 35, 35),
   'intercalation': False}),
 ({'date': (<Months.ELA: 9>, 19),
   'doy': 256,
   'preceding': (30, 30, 30, 30, 30, 29, 29, 29),
   'intercalation': False},
  {'date': (<Prytanies.VIII: 8>, 7),
   'doy': 256,
   'preceding': (36, 36, 36, 36, 35, 35, 35),
   'intercalation': False}))

This narrows the possibilities. To satisfy the equation, Elaphēboliṓn can only be preceded by 3–5 full months, prytany VIII by 2–4 long, 36-day prytanies and the year cannot be intercalary.

We can compare this to how `heniautos` calculates the year. It is important remember that `heniautos` only characterizes what the year _might have been_ if it followed a simplified version of astronomical observations. We can use the calendar equations to see how close `heniautos` comes to the historical evidence and use `heniautos` as a framework to help us see what the the equations tell us about the calendar.

For IG II³ 345 `equations` gave us possibilities from DOY 254 to DOY 256, with Elaphēboliṓn preceded by 3-5 full months and 2 to 4 long prytanies in a normal year. 

What does `heniautos` give for this year (332/1)?

In [18]:
for i, month in enumerate(ha.festival_calendar(ha.bce_as_negative(332)), 1):
    print(f"{i:>2} | {month['month']:12} | "
          f"{ha.as_eet(month['days'][0]['date'])} | {len(month['days'])}" )

 1 | Hekatombaiṓn | BCE 0332-Jul-21 | 30
 2 | Metageitniṓn | BCE 0332-Aug-20 | 30
 3 | Boēdromiṓn   | BCE 0332-Sep-19 | 29
 4 | Puanepsiṓn   | BCE 0332-Oct-18 | 30
 5 | Maimaktēriṓn | BCE 0332-Nov-17 | 29
 6 | Poseidēiṓn   | BCE 0332-Dec-16 | 30
 7 | Gamēliṓn     | BCE 0331-Jan-15 | 29
 8 | Anthestēriṓn | BCE 0331-Feb-13 | 30
 9 | Elaphēboliṓn | BCE 0331-Mar-15 | 29
10 | Mounuchiṓn   | BCE 0331-Apr-13 | 29
11 | Thargēliṓn   | BCE 0331-May-12 | 30
12 | Skirophoriṓn | BCE 0331-Jun-11 | 29


Based on the new moons for this year, `heniautos` generates an ordinary year, in which Elaphēboliṓn is indeed preceded by five full months (Hek, Met, Pua, Pos, and Anth), which matches the DOY 256 solution for IG II³ 345.

The conciliar calendar matches this solution as well with four 36-day prytanies preceding prytany VIII.

In [13]:
for i, pryt in enumerate(ha.prytany_calendar(ha.bce_as_negative(332)), 1):
    print(f"{i:>2} | {pryt['prytany']:>2} | "
          f"{ha.as_eet(pryt['days'][0]['date'])} | {len(pryt['days'])}")

 1 |  1 | BCE 0332-Jul-21 | 36
 2 |  2 | BCE 0332-Aug-26 | 36
 3 |  3 | BCE 0332-Oct-01 | 36
 4 |  4 | BCE 0332-Nov-06 | 36
 5 |  5 | BCE 0332-Dec-12 | 35
 6 |  6 | BCE 0331-Jan-16 | 35
 7 |  7 | BCE 0331-Feb-20 | 35
 8 |  8 | BCE 0331-Mar-27 | 35
 9 |  9 | BCE 0331-May-01 | 35
10 | 10 | BCE 0331-Jun-05 | 35


`heniautos` generates conciliar calendars that observe the Rule of Aristotle, placing all the long prytanies first. While this is likely to be true, the calendar equation supports any arrangement as long as four of the the first seven are 36 days long.

## Visualizing Calendar Equations

It may help to visualize the arrangment of full and hollow months, long and short prytanies so that we can diagram potential solutions to calendar equations. In the diagrams below, The first row of each pair represents the 12-month festival year, and the second the ten month conciliar year. full months/long prytanies are represented by darker, green boxes and hollow/short by lighter, beige boxes.  

The first diagram shows an "ideal" year according to most rigid conception: perfectly alternating full and hollow months, the conciliar year beginning with all 4 long prytanies. The second diagram shows a "schematic" view that is only concerned with the _counts_ of months rather than their arrangement. The festival year starts with five green boxes because there are at least five full months in the year, likewise for the next five beige boxes for the minimum count of five hollow months. The last two boxes are empty because we cannot, in fact, say what they are. If one is full and one hollow, we have an ideal 354-day year. But some years could be 353 days long (both hollow) and some 355 (both full). These last two months, schematically, are wildcards. The conciliar year has no wildcards because they were always 4 long and 6 short. The schematic representation of the conciliar year happens to follow the Rule of Aristotle. but that is just a coincidence.


<img src="examples/ideal.png" width="700">

We can diagram our results for IG II/III³ 1, 345 like this:

<img src="examples/ig-ii3-345-all.png" width="700">

The dotted line represents the horizon of the calendar equation, which tells us what it can about the first eight months of the year (preceding Elaphēboliṓn) and the first seven prytanies (preceding prytany VIII). The months before Elaphēboliṓn, to the left of the dotted line, are the months that must satisfy the equation. The months to the right of the dotted line, with hash marks, are the remaining, unknown months. We can deduce the lengths of some of the unknown months from the known. For instance any normal year has _at least_ five full months as well as five hollow months. In first diagram, DOY 253, only two of the known months are full so at least three of the unknown must be full. Since there are already six known hollow months and the maximum is sevem, the last month is a wildcard. The conciliar year must have four long prytanties; since there is only 1 long prytany to the left of the dotted line, the remaining three have to all be long. The diagrams make it easy to see that only one solution satisfies the Rule of Aristotle, DOY 256, if we believe that that is important.

Using these diagrams we can illustrate how one of these solutions matches an "ideal" festival year. Of the first eight months of an ideal year, four must be full and four hollow. Only one solution matches this, DOY 255. Remember, the schematic diagrams do not show an arrangement but only a count of months before and after the equation. We can rearrange the schematic months to match the ideal.

<img src="examples/ig-ii3-345-ideal.png" width="700">

For the solution to be a perfect match, it is important that no arrow cross the dotted line. Because `heniautos` calculates that _five_ of the first eight months were full, we cannot match the schematic diagram for DOY 255 to one representing `heniautos`' arrangement of months. With the wildcards we can make the counts work, but the arrow connecting the last full month on each will cross the line, indicating that it does not satisfy the equation:

<img src="examples/ig-ii3-345-doy-255-heniautos.png" width="700">

We can however match up `heniautos`' calculated year with the DOY 256 solution:

<img src="examples/ig-ii3-345-doy-256-heniautos.png" width="700">

With the aid of the daigrams we can see that if we insist on an ideal arrangement of alternating months, we can have *Elaphēboliṓn 19 = Prytany VIII 7 = DOY 255* as long as we give up the Rule of Aristotle. If we want the Rule of Aristotle we can have *Elaphēboliṓn 19 = Prytany VIII 7 = DOY 256*, which _also_ matches possible observations of the full moon. `heniautos` can give us its Julian date for this Elaphēboliṓn 19:

In [5]:
ha.as_eet(ha.find_date(ha.bce_as_negative(332), ha.Months.ELA, 19)["date"])

'BCE 0331-Apr-02'

So we can fill out an extended equation:

    Elaphēboliṓn 19 = Prytany VIII 7 = DOY 256 = April 2, 331 BCE

## Multiple Equations

[IG II/III³ 1, 348](http://telota.bbaw.de/ig/digitale-edition/inschrift/IG%20II_III%C2%B3%201,%20348), from the same year (ἐπὶ Νικήτου ἄρχοντος) has very clear dates:

    ἐπὶ Νικήτου ἄρχοντος, ἐπὶ τῆς Ἐρεχθη–
    ίδος ἐνάτης πρυτανείας, ἧι Ἀριστόνο–
    υς Ἀριστόνου Ἀναγυράσιος ἐγραμμάτ–
    ευεν· Θαργηλιῶνος ἑνδεκάτει· τρίτηι
    καὶ εἰκοστῆι τῆς πρυτανείας
    
and gives us this calendar equation:

    Thargēliṓn 11 = Prytany IX.23
    
which has three solutions:

<img src="examples/ig-ii3-348.png" width="700">

All the solutions require a normal year, so this is generally in agreement with our solutions for IG II³ 345. Can we combine the two equations more specifically? With 4 solutions for the first and 3 for the second there 12 possible combinations, but to be a valid combination the equation for the earlier part of the year must partially satisfy that for the later part of the year. Only 6 of the possible combinations are valid in this way (scripts for performing these calculations and creating these charts are in the `examples` directory of the repository)

<img src="examples/ig-ii3-345-348-combo.png" width="700">

As an example, DOY 255 which requires 4 full and 4 hollow months, can be combined with DOY 306 which requires 5 full and 5 hollow months. We "subtract" DOY 255 from DOY 306, so that together they are (4 full + 4 hollow ) + (1 full + 1 hollow). We can still match this combined solution with an "ideal" year of alternating months without crossing either dotted line.

<img src="examples/ig-ii3-345-348-doy-255-306-solution.png" width="700">

With the extra information from IG II/III³ 1, 348, though, our the `heniautos` solution is no longer adequate.

<img src="examples/ig-ii3-345-348-doy-256-307-solution.png" width="700">

With only one equation we had two hollow and two wildcard months to work with in the last four months. Now the the second equation requires either either the 9th or the 10th month to be full while `heniautos` calculates them both as hollow, we have arrows crossing the dotted lines. 

## Making Adjustments

`heniautos` cannot and does not aim to generate historically _accurate_ calendars, just historically _possible_ or maybe even _probable_ calendars if its simplified "observations" are close to what the Athenians might have made. It may serve as a framework to help make truly historical reconstructions where there is other evidence. What can we do with the two equations from IG II³ 345 and IG II³ 348?

Again, it is important to keep in mind that schematic representation communicates the _count_ not to exact _arrangement_ of months. There is no particular reason to put all the full months first and the hollow second. Within the bounds of the equations, we can move the schematic months around however we need to. If we switch the full and hollow 9th and 10th month, that brings us a little closer to `heniautos`' calculation. The count is the same: one full and one hollow in months 9 and 10.

<img src="examples/ig-ii3-345-348-doy-256-307-rearr.png" width="700">

Turning to `heniautos`, Mounuchiṓn is hollow and Thargēliṓn is full. The new moon at the beginning of Thargēliṓn is "observed" on May 12 but, as we have said before, these dates are probably accurate to within about a day. It is within this margin of error to make May 13 the beginning of Thargēliṓn. This change of one day makes Mounuchiṓn full and Thargēliṓn which "uncrosses" the lines on the diagram. This adjustment is represented on the diagram by the reciprocal arrows within the boxes of the 10th and 11th months.

<img src="examples/ig-ii3-345-348-doy-256-307-adj.png" width="700">

The adjustment has no effect outside these two months, and none of this has any effect on the prytanies. Mounuchiṓn still begins on April 13, and Thargēliṓn still ends on June 10. Nor does it affect the basic equation if we want to stay as close as possible to the "observed" new moons, and preserve the Rule of Aristotle

    Thargēliṓn 11 = Prytany IX.23 = DOY 306
    
Without the adjustment, DOY 306 would have been May 22.

In [8]:
date = ha.find_date(ha.bce_as_negative(332), ha.Months.THA, 11)
print("DOY:", date["doy"])
print("Date:", ha.as_eet(date["date"]))

DOY: 306
Date: BCE 0331-May-22


With the adjustment, since Thargēliṓn starts one day later, the equation will be:

    Thargēliṓn 11 = Prytany IX.23 = DOY 306 = May 23, 331 BCE 
    
If we do not think it is important to observe the Rule of Aristotle, then the equation that worked with the "ideal" year, DOY 355 + DOY 306, is also possible with one adjustment.

<img src="examples/ig-ii3-345-348-doy-255-306-rearr.png" width="700">

Here we have adjusted the first day of Elaphēboliṓn, moving it from March 15 to March 14 to make Anthestēriṓn hollow and Elaphēboliṓn full.

In [ ]:
ha.equations((ha.Months.THA, 11), (ha.Prytanies.IX, 23), pryt_type=ha.Prytany.ALIGNED_10)


[IG II/III³ 1, 331](http://telota.bbaw.de/ig/digitale-edition/inschrift/IG%20II_III%C2%B3%201,%20331) is more fragmentary, but what remains is enough for certainty:

    [ἐπὶ Εὐαι]ν̣έτου ἄρχοντος, ἐπὶ [τῆ]–
    [ς Ἀντι]οχίδος δεκάτης πρυτα[νε]–
    [ίας, ἧι] Πρόξενος Πυλαγόρου [Ἀχε]–
    [ρδούσ]ιος ἐγραμμάτευεν· Σ[κιρο]–
    [φορι]ῶνος ὀγδόηι ἐπὶ δέκα· τρ̣[ίτ]–
    [ηι κ]αὶ εἰκοστῆι τῆς πρυταν̣ε̣[ία]–
    [ς· 
    
The equation is

    Skirophoriṓn 18 = Prytany X.23

In [ ]:
ha.equations((ha.Months.SKI, 18), (ha.Prytanies.X, 23), pryt_type=ha.Prytany.ALIGNED_10)

The equation has only one valid "solution":

    Skirophoriṓn 18 = Prytany X.23 = DOY 342
    
This requires Skirophoriṓn to be preceded by 5 full and 6 hollow months--(5 × 30) + (6 × 29) + 18 = 342. The conciliar date needs to be preceded by 4 36-day and 5 35-day prytanoes--(4 × 36) + (5 × 35) + 23 = 342--meaning it is not an intercalary year. There are no other combinations that have the same DOY while also satisfying other requirements (for example, since this is the 10th prytany at least 3 and at most 4 long prytanies can preceded it)

In [ ]:
for i, month in enumerate(ha.festival_calendar(ha.bce_as_negative(332)), 1):
    print(f"{i:>2} | {month['month']:12} | {ha.as_eet(month['days'][0]['date'])} | {len(month['days'])}" )

IG II/III³ 1, 375
IG II/III³ 1, 376
IG II/III³ 1, 378

In [ ]:
ha.equations((ha.Months.PUA, 18), (ha.Prytanies.III, 36), pryt_type=ha.Prytany.ALIGNED_10)

In [ ]:
for i, month in enumerate(ha.festival_calendar(ha.bce_as_negative(323)), 1):
    print(f"{i:>2} | {month['month']:12} | {ha.as_eet(month['days'][0]['date'])} | {len(month['days'])}" )

The most concrete scholarship relating to the ancient Athenian calendar involves "calendar equations." Inscriptions that allow us to connect a festival date with a conciliar date give us clues to the nature of the calendar and can help us establish intercalations. `heniautos` has features that can help explore these calendar equations.

## [IG II/III³ 1, 1292](http://telota.bbaw.de/ig/digitale-edition/inschrift/IG%20II_III%C2%B3%201,%201292)

Pritchett and Neugebauer (1947) 16 use this inscription as evidence of their interpretation of dates κατ' ἄρχοντα "according to the árkhōn" and κατὰ θεὸν "according to the heavens".   

    ἐπὶ Χαρικλέους ἄρχοντος, ἐπὶ τῆς Αἰγεῖδος ἐνάτης πρυτανείας, ἧι
    Αἰσχρίων Εὐαινέτου Ῥαμνούσιος ἐγραμμάτευεν· δήμου ψηφίσματα·
    Ἐλαφηβολιῶνος τρίτει ἐπὶ δέκα, κατὰ θεὸν δὲ ὀγδόει <ἐπὶ δέκα, ὀγδόει> καὶ εἰκοστεῖ
    τῆς πρυτανείας· 

> In the arkhonship of Khariklēs, in the ninth prytany of Aigēis, for which Aiskhríōn son of Euaínetos of Rhamnoûs was secretary; decrees of the people; Elaphēboliṓn 13, 18 according to the heavens, 28th of the prytany.

There is one prytany date here, IX 28 (as prytany dates are often written), and two festival dates Elaphēboliṓn 13 or, "according to the heavens," Elaphēboliṓn 18. Pritchett and Neugebauer wish to show that the κατὰ θεὸν dates are dates according to the "natural" lunisolar calendar while the κατ' ἄρχοντα dates have been adjusted in some way that is out of alignment with the natural calendar. They work out their calendar equation in this way (the "_Hesperia_ text" is Meritt (1936) 419-428 and the reference to Dinsmoor is Dinsmoor \[1939\]):

> Prytany IX 28 in an intercalary year would correspond to the 284th day. This day, according to the _Hesperia_ text, is to be equated with κ.θ. IX 18; thus κ.θ. IX 1 = 284 - 17 = 267th day of the year, beginning with Hekatombainon 1. Accepting Dinsmoor's Julian date 196 B.C. June 18 we obtain for the 267th day 195 B.C. March 11. This day is one day later that the astronomical new moon; thus the first day of a month κατὰ θεὸν appears to be exactly what one would expect for the first day of a lunar calendar

Pritchett and Neugebauer skip a number of steps in their explanation, beginning with their acceptance that the year 196/195 ("arkhonship of Khariklēs") is intercalated, then working backwards from Elaphēboliṓn 18 to identify Elaphēboliṓn 1 (267th day) as the day after an astronomical new moon. `heniautos` can help us work out all the steps more explicitly.

## Finding the DOY (Day of the Year)

"Day of year" just means what number a day or date is since the beginning of that year. For example, on the Gregorian (and Julian) calendar February 1, coming after January 31, is always the 32nd day of the year. March 1 is usually the DOY 60 (preceded by 31 + 28 = 59 days), but in a leap year it is DOY 61 (preceded by 31 + 29 = 60 days). Every date _after_ February can be one of two days of the year depending on whether the year is a leap year or not.

This is more complicated on the Athenian calendar. Metageitniṓn 1 (the first day of the second month) is DOY 30 if Hekatombaiṓn is "hollow" (29 days) but DOY 31 if Hekatombaiṓn is full. Boēdromiṓn 1 can be DOY 59, 60, or 61 depending on whether it is preceded by two hollow, one hollow and one full (in either order), or two full months.

Then there is intercalation. Metageitniṓn _could be_ the _third_ month  if a month is intercalated before it. Then Metageitniṓn 1 would be DOY 59, 60, or 61 depending on the lengths of the preceding months. Boēdromiṓn would then be the 4th month and Boēdromiṓn 1 could be any of DOY 88-91.

`heniautos` provides a function, `festival_doy()` that will calculate all the possible days of the year a Athenian date might have and what conditions must precede it. It takes constant from `ha.Months` and a day.

In [ ]:
import heniautos as ha
ha.init_data()

ha.festival_doy(ha.Months.MET, 1)

These results mean that Met 1 is DOY 30 when it is preceded by 0 30-day months and 1 29-day. `'intercalated': False` means that none of the preceding months are an intercalation. Met 1 is DOY 60 when it preceded by 1 30-day and 1 29-day month, and one of those months _has to be an intercalation_ (`'intercalated': True`). Note that `'intercalated': False` does not mean that there is no intercalation in the year, but only none before the particular date. One could follow, but it does not affect the DOY calculation.

In actuality, DOY 59 (Met 1 preceded by two hollow months) is almost certainly ruled out. We would expect an intercalated month to be 30 days, so if Metageitniṓn is preceded by an intercalation, one of them _should_ be full. `heniautos` filters out many mathematically possible but practically impossible results, but it does err on the side of generosity, as with Met 1 = DOY 59.

The trick with calendar equations is to find a date in the festival calendar that is the same DOY as a date on the conciliar year. In the above example we have two dates in Elaphēboliṓn, the 13th and the 18th to work with. What are the possible DOY values for Elaphēboliṓn 13?

In [ ]:
ha.festival_doy(ha.Months.ELA, 13)

The first option, DOY 247, can only happen if Elaphēboliṓn is preceded by two full and six hollow months ((30 × 2) + (29 × 6) + 13 =  247). Since Elaphēboliṓn is normally the 9th month, eight preceding months means there can have been no intercalation in this count (`'intercalated': False`). Again, this does not mean there is no intercalation _in the year_ since the intercalation could _follow_ Elaphēboliṓn. 

The last option, DOY 280, comes from six full and 3 hollow months preceding ((30 × 6) + (29 × 3) + 13 =  280). Since there are 9 preceding months, making Elaphēboliṓn the 10th month in this case, one of the preceding months must be an intercalation (`'intercalated': True`).

What about Elaphēboliṓn 18?

In [ ]:
ha.festival_doy(ha.Months.ELA, 18)

So we have days with ranges within which the dates must fall: Ela 13 = DOY 247-256 and Ela 18 = DOY 282-285.

Now we come to the conciliar year. What are the possible DOY values for Prytany IX 28? `festival_doy()` has a companion, `prytany_doy()` for calculating this. It takes a number for the prytany, a number for the day, and a constant from `heniautos.Prytany` for the kind of conciliar year we expect. In 196/195 BCE there were twelve tribes and therefore 12 prytanies and the conciliar year began and ended with the festival year. In `heniautos` terms we want to use `Prytany.ALIGNED_12`.

In [ ]:
ha.prytany_doy(ha.Prytanies.IX, 28, pryt_type=ha.Prytany.ALIGNED_12)

The result returned by `prytany_doy()` is similar, but intercalation here does mean that the _year_ is intercalated becuase that changes the _lengths_ of the preceding prytanies rather than their _number_ as it does the number of months of the festival calendar. In this period an intercalated year would have twelve 32-day prytanies, while a normal year would have twelve 30- or 29-day prytanies following the lunar months. For this example, all the possible values for the DOY involve eight preceding prytanies. Either they were all 32 days long (DOY 284) or they were some combination of 29 and 30 days (DOY 260-268). 

Some of these values (such as 8 29-day prytanies) are no more practically possible than they are for the festival calendar but, because there are so many variations of the conciliar calendar along with the unresolved question of the "Rule of Aristotle" that `prytany_doy()` does not try to exclude these.

Scanning these values will tell us that there is only one possible calendar equation, but `heniautos` provides a function to do it for us. This will find matching options, if they exist:

In [ ]:
ha.equations((ha.Months.ELA, 13), (ha.Prytanies.IX, 28), pryt_type=ha.Prytany.ALIGNED_12)

In [ ]:
ha.equations((ha.Months.ELA, 18), (ha.Prytanies.IX, 28), pryt_type=ha.Prytany.ALIGNED_12)

`equations()` takes a tuple for the festival date (an `ha.Months` constant and an integer for the day) and a tuple for the prytany date (an `ha.Prytanies` constant and the day). In the examples above we have also specified a `pryt_type` as with `prytany_doy()`.

This confirms Pritchett and Neugebauer. There is no solution that satisfies the equation:

    Elaphēboliṓn 13 = Prytany IX 28
    
That is, there is no possible DOY value for this festival calendar date that matches any possible DOY values for this conciliar year date. For Elaphēboliṓn 18, though, there is exactly one solution

    Elaphēboliṓn 18 = Prytany IX 28 = DOY 284
    
The κατὰ θεὸν date does match a "natural" date of the lunisolar calendar. The κατ' ἄρχοντα date must indicate an earlier manipulation of the calendar (for what reason we can't say) that caused the "official" Elaphēboliṓn to start 5 days later than the astronomical.

Furthermore, our solution specifies that it be an intercalated year, that the intercalation occur before Elaphēboliṓn, and that Elaphēboliṓn be preceded by 5 full and 4 hollow months. How does this compare to the festival year as calculated by `heniautos`? Meritt (1936) and Pritchett and Neugebauer (1947) date the arkhonship of Khariklēs to 196/195 BCE:

In [ ]:
for i, month in enumerate(ha.festival_calendar(ha.bce_as_negative(196)), 1):
    print(f"{i:>2} | {month['month']:12} | {ha.as_eet(month['days'][0]['date'])} | {len(month['days'])}" )

Not so good. According to `heniautos`' calculations, 196/195 is _not_ an intercalary year and cannot satisfy any calendar equation where Prytany IX 28 = DOY 284. There are three main possiblities: 1) `heniautos`' calculations are wrong, 2) `heniautos` calculations are "right" but do not match the historical calculations or manipulations made by the Athenians of the 2nd century BCE, 3) the year is wrong.

In this case, actually, the year is not so certain. Khariklēs is now generally dated to 184/183:

In [ ]:
for i, month in enumerate(ha.festival_calendar(ha.bce_as_negative(184)), 1):
    print(f"{i:>2} | {month['month']:19} | {ha.as_eet(month['days'][0]['date'])} | {len(month['days'])}" )

Better at first glance since we have an intercalary year. However, this Elaphēboliṓn is preceded by 4 full and 5 hollow months and we were expecting the reverse, 5 full and 4 hollow. If we look at Elaphēboliṓn itself, according to `heniautos`, the 18th is DOY 283 not 284.

In [ ]:
for d in ha.festival_calendar(ha.bce_as_negative(184))[9]["days"]:
    print(f"Ela {d['day']:>2} | {ha.as_eet(d['date'])} | {d['doy']}")

Still it is only one day's difference. Given the uncertainty in how well these astronomically-based calculations match with the real, historical observations we might satisfy ourselves that this is close enough.

There is still one more possibility, though. The date of Khariklēs' arkhonship is not certain. Knoepfler (2015) would place it instead in 200/199. The editors of AIO (n.d.) give both dates for this arkhonship because "At present there seem to be no decisive arguments and we therefore leave both possibilities open."

What does `heniautos` say for 200/199 BCE?

In [ ]:
for i, month in enumerate(ha.festival_calendar(ha.bce_as_negative(200)), 1):
    print(f"{i:>2} | {month['month']:19} | {ha.as_eet(month['days'][0]['date'])} | {len(month['days'])}" )

We have an intercalary year and this Elaphēboliṓn is preceded by 5 full and 4 hollow months, exactly what we are looking for. This Elaphēboliṓn 18 is DOY 284

In [ ]:
for d in ha.festival_calendar(ha.bce_as_negative(200))[9]["days"]:
    print(f"Ela {d['day']:>2} | {ha.as_eet(d['date'])} | {d['doy']}")

Now we have an exact calendar equation:

    Elaphēboliṓn 18 = Prytany IX 28 = DOY 284 = BCE 0199-Apr-11
    
I would not go as far as saying that this proves that Khariklēs was arkhon in 200/199 but it is an interesting bit of evidence.

## `festival_doy()` and `prytany_doy()` in Detail

As mentioned above, `festival_doy()` excluded less probable DOY values based on the mechanics of the lunar cycles. `heniautos` rarely generates 4 months of the same length, either hollow or full, in row and does not seem to ever generate 5. Given five months, the maximum difference between the number of each will be 3 (e.g. FFFFh = a difference of 3). With 6 months sequences like FFFFhF, with a difference of 4, are possible but no greater difference seems to occur.

The default behavior, then, is for `festival_doy()` to exclude any combinations of full and hollow months that differ in number by greater than 4. You can choose your own values with the `max_diff` parameter. For example `festival_doy(ha.Months.ELA, 18, max_diff=3)` will return fewer possibilities. Use `festival_doy(ha.Months.ELA, 18, max_diff=0)` to do no filtering at all and return all possibilities.

`prytany_doy()` does not do any filtering by default (that is, its default is `max_diff=0`). You can supply `max_diff` if you want.

If you give `prytany_doy()` a year and do not specify a `Prytany` constant (see [Choosing-a-Prytany-Variation](Heniautos%20Basics.ipynb#Choosing-a-Prytany-Variation)) it will choose the prytany type based on the year: 


In [ ]:
ha.prytany_doy(ha.Prytanies.IX, 28, ha.bce_as_negative(200))

Otherwise you need to specify the prytany type in the examples above. This is the equivalent of the previous (but you could give a different prytany type of you thought the automatic choice was wrong):

In [ ]:
ha.prytany_doy(ha.Prytanies.IX, 28, pryt_type=ha.Prytany.ALIGNED_12)

## Works Cited

* Dinsmoor, William Bell. 1939. _The Athenian Archon List in the Light of Recent Discoveries_. New York: Columbia University Press. https://doi.org/10.7312/dins93096.
* Knoepfler, D. 2015. “Τὰ τριάκοντα ἔτη. Observations sur le texte, le contenu et la date du décret d’Athènes pour Képhisodoros.” In _Α‬ξων: studies in honor of Ronald S. Stroud_, edited by Angelos P Matthaiou and Nikolaos Papazarkadas, 257–90. Hellēnikē Epigraphikē Hetaireia 17.
* Meritt, Benjamin D. 1936. “Greek Inscriptions.” _Hesperia_ 5 (3): 355–430. https://doi.org/10.2307/146623.
* Pritchett, W. Kendrick, and O. Neugebauer. 1947. _The Calendars of Athens_. Cambridge: Harvard University Press.
* AIO. n.d. “IGII31 1292 Honours for Kephisodoros.” Accessed April 3, 2021. https://www.atticinscriptions.com/inscription/IGII31/1292.

In [ ]:
from itertools import zip_longest

months = tuple(zip_longest((ha.Months.ELA,), range(11,20), fillvalue=(ha.Months.ELA)))
pryts = tuple(zip_longest((ha.Prytanies.VIII,), range(1,11), fillvalue=(ha.Months.ELA)))
eq = ha.equations(months, pryts, pryt_type=ha.Prytany.ALIGNED_10)

for e in eq:
    for f in zip_longest((e["doy"],), e["equations"]["festival"], e["equations"]["conciliar"], fillvalue=False):
        print(f"{f[0] if f[0] else ''} | {f[1]['date'] if f[1] else ''} | {f[2]['date'] if f[2] else}")
        



In [ ]:
#IG II2 546

#zip_longest(range(11,20)
pryts = [a for b in [tuple(zip_longest((p,), tuple(range(11,20)), fillvalue=p)) 
                     for p in (ha.Prytanies.VII, ha.Prytanies.VIII, ha.Prytanies.IX)] for a in b]
ha.equations((ha.Months.ANT, 21), pryts, pryt_type=ha.Prytany.ALIGNED_10)